<a href="https://colab.research.google.com/github/del-b2l/sem-6-assignments/blob/main/Assignment_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 1: Neural Network and Backpropagation from Scratch

**Course:** Deep Learning for Perception  
**Semester:** Spring 2026  

**Student Name:**  Syeda Batool Kazmi  
**Student ID:**  23K-0672

---

## Instructions
- Do NOT use deep learning libraries (TensorFlow, PyTorch, Keras).
- You may use `numpy` and `math`.
- Follow the steps carefully.
- Show intermediate values clearly.
- Do not change the given architecture.
- Compute values of x1 and x2 from your ID. For instance if the ID is Student ID: 23K-1365

Digits → 1, 3, 6, 5

x1=0.13

x2=0.65

In [ ]:
import numpy as np
import math


## Problem Description

We implement a small feedforward neural network with:
- 2 input neurons
- 1 hidden layer with 2 neurons (sigmoid activation)
- 1 output layer with 2 neurons (sigmoid activation)

We will:
1. Perform forward propagation
2. Compute Mean Squared Error
3. Apply backpropagation
4. Update weights using gradient descent


In [ ]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

def sigmoid_derivative(z):
    s = sigmoid(z)
    return s * (1 - s)


In [ ]:
# Inputs
x1 = 0.06
x2 = 0.72

# Targets
t1 = 0.01
t2 = 0.99

# Weights (Input → Hidden)
w1, w2 = 0.15, 0.20
w3, w4 = 0.25, 0.30

# Weights (Hidden → Output)
w5, w6 = 0.40, 0.45
w7, w8 = 0.50, 0.55

# Biases
b1 = 0.35
b2 = 0.60

# Learning rate
eta = 0.5


## Forward Propagation

Compute activations of the hidden layer and output layer.


In [ ]:
# Hidden layer
z_h1 = x1*w1 + x2*w2 + b1
z_h2 = x1*w3 + x2*w4 + b1

a_h1 = sigmoid(z_h1)
a_h2 = sigmoid(z_h2)

# Output layer
z_o1 = a_h1*w5 + a_h2*w6 + b2
z_o2 = a_h1*w7 + a_h2*w8 + b2

y1 = sigmoid(z_o1)
y2 = sigmoid(z_o2)

print("Hidden activations:", a_h1, a_h2)
print("Output predictions:", y1, y2)


Hidden activations: 0.6231640828996309 0.6412974738474874
Output predictions: 0.7572845851894309 0.7797646939567906


## Loss Computation

We use Mean Squared Error (MSE).


In [ ]:
E1 = 0.5 * (t1 - y1)**2
E2 = 0.5 * (t2 - y2)**2
E_total = E1 + E2

print("Total Error:", E_total)


Total Error: 0.30131656758441083


## Backpropagation

Compute gradients using the chain rule.


In [ ]:
# Output layer gradients
dE_dy1 = y1 - t1
dE_dy2 = y2 - t2

dy1_dz1 = sigmoid_derivative(z_o1)
dy2_dz2 = sigmoid_derivative(z_o2)

# Gradients for weights
dE_dw5 = dE_dy1 * dy1_dz1 * a_h1
dE_dw6 = dE_dy1 * dy1_dz1 * a_h2
dE_dw7 = dE_dy2 * dy2_dz2 * a_h1
dE_dw8 = dE_dy2 * dy2_dz2 * a_h2

print("Gradients:", dE_dw5, dE_dw6, dE_dw7, dE_dw8)


Gradients: 0.08559431364023416 0.0880850142353803 -0.02249875959184728 -0.02315344880567752


## Weight Update

Update weights using gradient descent.


In [ ]:
w5 -= eta * dE_dw5
w6 -= eta * dE_dw6
w7 -= eta * dE_dw7
w8 -= eta * dE_dw8

# for reference
print("Old weights:")
print(0.40, 0.45, 0.50, 0.55)

print()

print("Updated weights:")
print(w5, w6, w7, w8)



Old weights:
0.4 0.45 0.5 0.55

Updated weights:
0.35720284317988293 0.40595749288230987 0.5112493797959237 0.5615767244028388


## Testing Updates to Code

In [ ]:
# checking Total Error after updating weights

print("Total Error (old):", E_total)

# Hidden layer
z_h1 = x1*w1 + x2*w2 + b1
z_h2 = x1*w3 + x2*w4 + b1

a_h1 = sigmoid(z_h1)
a_h2 = sigmoid(z_h2)

# Output layer computation with the new weights in the previous code block
z_o1 = a_h1*w5 + a_h2*w6 + b2
z_o2 = a_h1*w7 + a_h2*w8 + b2

y1 = sigmoid(z_o1)
y2 = sigmoid(z_o2)

print("Hidden activations:", a_h1, a_h2)
print("Output predictions:", y1, y2)

E1 = 0.5 * (t1 - y1)**2
E2 = 0.5 * (t2 - y2)**2
E_total = E1 + E2

print("Total Error (new):", E_total)

Total Error (old): 0.30131656758441083
Hidden activations: 0.6231640828996309 0.6412974738474874
Output predictions: 0.7470490967957893 0.7822335143314326
Total Error (new): 0.29320414182727794


In [ ]:
# for reference
print("Old weights:")
print(w5, w6, w7, w8)

print()

# assigning a new learning rate
eta = 1.5

w5 -= eta * dE_dw5
w6 -= eta * dE_dw6
w7 -= eta * dE_dw7
w8 -= eta * dE_dw8

print("Updated weights:")
print(w5, w6, w7, w8)

Old weights:
0.35720284317988293 0.40595749288230987 0.5112493797959237 0.5615767244028388

Updated weights:
0.2288113727195317 0.2738299715292394 0.5449975191836945 0.596306897611355


## Some more tasks

1. Did the total error decrease after one iteration?
2. What would happen if the learning rate were too large?
3. Repeat this with a larger learning rate and one more input. You may select x3 with a random value between .1 and .5


## Answers

1. Yes, the **total error decreased** after one iteration from a value of 0.3013 to 0.2932. This indicates that our gradient descent algorithm is reducing error as expected, while our system is learning.
2. Although a larger learning rate can speed up training, if it is too large it **may cause the weight updates to overshoot the minimum**. This can lead to **instability**, **oscillations in the error**, or **failure to converge**.
3. **Exercise repeated below** with eta = 2.5 and x3 = 1.6.

In [ ]:
# Inputs
x1 = 0.06
x2 = 0.72
# New input variable x3
x3 = 0.16

# Targets
t1 = 0.01
t2 = 0.99

# Weights (Input → Hidden)
w1, w2, w3 = 0.15, 0.20, 0.25
w4, w5, w6 = 0.30, 0.40, 0.45

# Weights (Hidden → Output)
w7, w8 = 0.50, 0.55
w9, w10 = 0.60, 0.65

# Biases
b1 = 0.35
b2 = 0.60

# New larger learning rate
eta = 2.5

In [ ]:
# Hidden layer
z_h1 = x1*w1 + x2*w2 + x3*w3 + b1
z_h2 = x1*w4 + x2*w5 + x3*w6 + b1

a_h1 = sigmoid(z_h1)
a_h2 = sigmoid(z_h2)

# Output layer
z_o1 = a_h1*w7 + a_h2*w8 + b2
z_o2 = a_h1*w9 + a_h2*w10 + b2

y1 = sigmoid(z_o1)
y2 = sigmoid(z_o2)

print("Hidden activations:", a_h1, a_h2)
print("Output predictions:", y1, y2)

Hidden activations: 0.6325100178154341 0.6743662330024384
Output predictions: 0.7836654977706325 0.805000210862238


In [ ]:
E1 = 0.5 * (t1 - y1)**2
E2 = 0.5 * (t2 - y2)**2
E_total = E1 + E2

print("Total Error:", E_total)

Total Error: 0.3163916122108485


In [ ]:
# Output layer gradients
dE_dy1 = y1 - t1
dE_dy2 = y2 - t2

dy1_dz1 = sigmoid_derivative(z_o1)
dy2_dz2 = sigmoid_derivative(z_o2)

# Gradients for weights
dE_dw7 = dE_dy1 * dy1_dz1 * a_h1
dE_dw8 = dE_dy1 * dy1_dz1 * a_h2
dE_dw9 = dE_dy2 * dy2_dz2 * a_h1
dE_dw10 = dE_dy2 * dy2_dz2 * a_h2

print("Gradients:", dE_dw7, dE_dw8, dE_dw9, dE_dw10)

Gradients: 0.08296160648132642 0.08845157305155585 -0.018368292121399978 -0.019583809925064893


In [ ]:
# for reference
print("Old weights:")
print(w7, w8, w9, w10)

w7 -= eta * dE_dw7
w8 -= eta * dE_dw8
w9 -= eta * dE_dw9
w10 -= eta * dE_dw10

print()

print("Updated weights:")
print(w7, w8, w9, w10)

Old weights:
0.5 0.55 0.6 0.65

Updated weights:
0.29259598379668394 0.32887106737111044 0.6459207303035 0.6989595248126622
